### DISCUSSION TITLE
Adding declarative recipes to a Recipe#getRecipeList

### DISCUSSION THREAD

**Discussion by crankydillo** [openrewrite/rewrite#4351] [Q&A] [answered]
For various reasons, I need to use an imperative recipe that ultimately returns a `getRecipeList`.  That recipe list needs to include declarative recipes that are found on the classpath.  Any patterns for this?  I'm currently using `Environment.Builder#scanYamlResources` to build a name->Recipe map from within my composite Recipe's constructor, but it just doesn't feel right..

**Comment by timtebeek** [comment, marked as answer]
Do I understand correctly that these recipes exclusively exist in local declarative recipes? As in they are not references to regular recipes with a constructor you might call instead?

Then indeed the workaround through `Environment` that you've listed might be the only good way to go about this.

I don't see this being any easier using the underlying `YamlResourceLoader` or `DeclarativeRecipe` classes themselves. You might be interested to learn of this new effort, but unless we add a declarative look up there it's not of much use to you here.
- https://github.com/openrewrite/rewrite/pull/4331

And please note you're also welcome to ask questions on [our Slack](https://join.slack.com/t/rewriteoss/shared_invite/zt-nj42n3ea-b~62rIHzb3Vo0E1APKCXEA); that typically gets a faster response.

**Reply by crankydillo** [reply]
They may be local declarative ones, but not necessarily.  The first declarative recipe that I need to add to my imperative Recipe's `#getRecipeList` is `org.openrewrite.apache.commons.collections.UpgradeApacheCommonsCollections_3_4`.  I added `rewrite-apache` to project dependencies.

**Reply by timtebeek** [reply]
ah yes, then that could make sense; what's the reason your own recipe has to be an imperative recipe? Could that be solved with a precondition on a declarative recipe?

**Reply by crankydillo** [reply]
The initial reason was to test the composite recipe.  Testing composite, as opposed to individual, may not be norm, but we are pretty excessive with our testing.  I did briefly look into modifying the YAML somehow as part of the test.  However, even if that worked, I'm thinking what we're doing with the imperative route is going to work well for multiple reasons.

**Reply by timtebeek** [reply]
Do know that you can also test declarative recipes, [as seen here](https://github.com/openrewrite/rewrite-apache/blob/75ca72cb8cb840fdd7f261b61d9e8f8709f64736/src/test/java/org/openrewrite/apache/commons/collections/UpgradeApacheCommonsCollections_3_4Test.java#L33-L37) for the Apache commons collections recipe you mentioned. I don't think there was any doubt to that, but wanted to call it out just to be sure.

Curious to hear what kind of recipes you're working on; let me know if there's anything we could potentially help adopt & distribute.

**Reply by crankydillo** [reply]
We'll definitely be opensource/contribute what we can!
